### PR TITLE
[SNAP-2975] fix SEGV in putInto

### PR DESCRIPTION
**Pull request by sumwale**
Issue was incorrect number of columns being determined in ColumnTableScan for the base table
due to output projection not having the extra virtual columns projected.

## Changes proposed in this pull request

- use relation schema to determine put/update case rather than output projection in ColumnTableScan
- removed duplicate methods across SharedUtils and Utils

## Patch testing

precheckin

## ReleaseNotes.txt changes

NA

## Other PRs 

NA

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala [hunk 1]
```diff
@@ -496,8 +496,10 @@ private[sql] final case class ColumnTableScan(
     val statsRow = ctx.freshName("statsRow")
     val deltaStatsRow = ctx.freshName("deltaStatsRow")
     val colNextBytes = ctx.freshName("colNextBytes")
-    val numTableColumns = if (ordinalIdTerm eq null) relationSchema.size
-    else relationSchema.size - ColumnDelta.mutableKeyNames.length // for update/delete
+    val numTableColumns = if (relationSchema.exists(
```
Comment (suranjan) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala]: How is it different?

Comment (dshirish) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala]: Why was the previous one wrong. I mean the ordinalId term check?
What we understood is that perhaps ordinalIDTerm == null meant no upate/delete has happened on it so just take the relationSchema.size else subtract the virtual columns.

Also why did you change size to length? Any issue there?

Please explain how was it wrong earlier?

Comment (sumwale) [reply core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala]: ordinalIDTerm will be set based on output, so if the output projection does not contain the extra columns then this check will go wrong. Specifically since putInto uses the same plan object for join/anti-join, there is a scan there which goes into the joins and is not fed to ColumnUpdateExec directly but still has those additional columns.

**Comment by sumwale** [comment]
@dshirish @suranjan As explained in the comments above, output projection may not have the extra columns used for update/delete with putInto. Specifically since putInto plan uses/combines same plan objects for join/anti-join one of the scans has the additional columns but does not get fed directly to ColumnUpdateExec rather is used only in the join, so output projection does not have those columns but the relation's schema (created from ColumnFormatRelation) has them.
